### PR TITLE
[SPARK-21052][SQL] Add hash map metrics to join

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

This adds the average hash map probe metrics to join operator such as `BroadcastHashJoin` and `ShuffledHashJoin`.

This PR adds the API to `HashedRelation` to get average hash map probe.

## How was this patch tested?

Related test cases are added.

### CODE REVIEW

**Comment by rxin** [comment]
Can you put a screenshot of the UI up, for both join and aggregate?

**Comment by viirya** [comment]
The screenshot of BroadcastHashJoin:

screen shot 2017-06-15 at 11 17 02 am

**Comment by viirya** [comment]
The screenshot of HashAggregate:

screen shot 2017-06-15 at 11 18 21 am

**Comment by viirya** [comment]
The screenshot of ShuffledHashJoin:

screen shot 2017-06-15 at 11 23 26 am

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 1]
```diff
@@ -573,8 +586,11 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
   private def updateIndex(key: Long, address: Long): Unit = {
     var pos = firstSlot(key)
     assert(numKeys < array.length / 2)
+    numKeyLookups += 1
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: you should also add this code to the het and the getValue methods.

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: Should we? It seems to me that we should only care about the hash collision happened when inserting the data into the hash map.

Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: IMO we should. The number of required probes is the different per key, and is also dependent on the order in which the map was constructed. If you combine this with some skew and missing keys, the number of probes can be much higher than expected.

You could even argue that we do not really care about the number of probes when building the map.

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: Yeah. OK. I think you're right. We should also care about the collision when searching keys in join operator. I'll update this in next commit.

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: Ain't you on a beach somewhere?!

**Comment by rxin** [comment]
I'd shorten it to "avg hash probe". Also do we really need min, med, max? Maybe just a single global avg?

**Comment by viirya** [comment]
So just get the global average of all avg hash probe metrics of all tasks? If there's skew, won't we like to see min, med, max?

**Comment by rxin** [comment]
yes but i just feel it is getting very long and verbose ..

**Comment by rxin** [comment]
also the avg probe probably shouldn't be an integer. at least we should show something like 1.9?

**Comment by viirya** [comment]
Because `SQLMetric` just stores long value. I was using a trick to multiply the avg probe by 1000 to get a long.

When preparing the values for UI, dividing the long with 1000 to get a float back.

So it's a workaround for long-based `SQLMetric`. But I finally don't use it.

Doesn't it sound too hacky for you?

**Comment by viirya** [comment]
Maybe just min and max?

**Comment by viirya** [comment]
The screenshot of UI showing float numbers instead of integer numbers.

screen shot 2017-06-15 at 4 27 38 pm

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala [hunk 2]
```diff
@@ -57,6 +57,11 @@ class SQLMetric(val metricType: String, initValue: Long = 0L) extends Accumulato
 
   override def add(v: Long): Unit = _value += v
 
+  // Because `SQLMetric` only stores long value, in order to store double average metrics, we
+  // multiply the given double with a base integer. When showing the metrics, it will be
+  // divided by the base integer to restore back a double.
+  def setWithDouble(v: Double): Unit = _value = (v * SQLMetrics.baseForAvgMetric).toLong
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: To store a float value into `SQLMetric`, currently I have no better idea to do. Any suggestion welcome.

**Comment by viirya** [comment]
cc @cloud-fan @gatorsmile for review.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala [hunk 3]
```diff
@@ -74,6 +80,19 @@ object SQLMetrics {
   private val TIMING_METRIC = "timing"
   private val AVERAGE_METRIC = "average"
 
+  private val baseForAvgMetric: Int = 10
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: I'm not quite sure this hack worth. For small number of probes, we don't care the values I think. For large number of probes, having one more digit in the fraction part is not very useful.

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: Yeah, that's why I record the ceil of average number at the beginning. cc @rxin What do you think?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 4]
```diff
@@ -422,8 +422,7 @@ class TungstenAggregationIterator(
         metrics.incPeakExecutionMemory(maxMemory)
 
         // Update average hashmap probe if this is the last record.
-        val averageProbes = hashMap.getAverageProbesPerLookup()
-        avgHashmapProbe.add(averageProbes.ceil.toLong)
+        avgHashmapProbe.set(hashMap.getAverageProbesPerLookup())
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: nit. What about changing `avgHashmapProbe` to `avgHashProbe` here, too?

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: Ah, I miss this one. Thanks @dongjoon-hyun

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala (hunk 1 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: Thanks for review even you're on a beach.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala (hunk 3 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: @cloud-fan @rxin Should I keep this way to show float metrics or fallback to ceil values?

**Comment by viirya** [comment]
ping @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: let's do ceil first.

**Comment by cloud-fan** [comment]
just realized that, for whole stage HashAggregate, we update the avgHashProb at last, but for non-whole-stage HashAggregate, we update it for every row. Shall we make it consistent?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala [hunk 5]
```diff
@@ -214,6 +215,7 @@ trait HashJoin {
     val resultProj = createResultProjection
     joinedIter.map { r =>
       numOutputRows += 1
+      avgHashProbe.set(hashed.getAverageProbesPerLookup())
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: why update it for every row? In HashAggregate we only update it at the end.

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: Because in HashAggregate, we process all rows in aggregation, then we can update the probe. But in join, it processes the rows as an iterator, we have no a point to update the probe in the end of join like aggregation case.

**Comment by viirya** [comment]
non-whole-stage HashAggregate is performed by `TungstenAggregationIterator`. The update to hash probe is only set at the end of the iterator. So I think it is consistent with whole stage HashAggregate?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 6]
```diff
@@ -422,8 +422,7 @@ class TungstenAggregationIterator(
         metrics.incPeakExecutionMemory(maxMemory)
 
         // Update average hashmap probe if this is the last record.
-        val averageProbes = hashMap.getAverageProbesPerLookup()
-        avgHashmapProbe.add(averageProbes.ceil.toLong)
+        avgHashProbe.set(hashMap.getAverageProbesPerLookup())
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: The probe for non-whole-stage HashAggregate is updated here. Actually, I now think we should move this update to the end of processing input rows, instead of at the last record. Because even the iterator is not consumed to the end (e.g., a Limit after Aggregate), we still should show the metric.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 7]
```diff
@@ -367,6 +367,10 @@ class TungstenAggregationIterator(
     }
   }
 
+  // Updating average hashmap probe after processing input rows. So even the iterator of result
+  // rows is not consumed to the last record, we still can show the metric.
+  avgHashProbe.set(hashMap.getAverageProbesPerLookup())
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: shall we also update `peakExecutionMemory` here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala (hunk 5 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: how about `TaskContext.addTaskCompletionListener`?

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: Ok. I also use TaskCompletionListener to update hash probe at the end of task for  non-whole-stage Join.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala (hunk 7 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: hmm, the peak memory of hashMap should not be changed during iterating the rows. But looks like the memory usage of the external sorter will change.  It might be meaningful to update `peakExecutionMemory` at the last record.

Maybe it is also good to use task completion listener to update it?

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: yea I think so, as the last record may not always be reached.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala [hunk 8]
```diff
@@ -91,6 +94,23 @@ case class BroadcastHashJoinExec(
...
 
   /**
+   * Returns the codes used to add a task completion listener to update avg hash probe
+   * at the end of the task.
+   */
+  private def genTaskListener(avgHashProbe: String, relationTerm: String): String = {
+    val listenerClass = classOf[TaskCompletionListener].getName
+    val taskContextClass = classOf[TaskContext].getName
+    s"""
+       | $taskContextClass$$.MODULE$$.get().addTaskCompletionListener(new $listenerClass() {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala]: maybe we can follow `incPeakExecutionMemory` and put it in `BufferedRowIterator`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 9]
```diff
@@ -686,6 +712,8 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
     writeLong(maxKey)
     writeLong(numKeys)
     writeLong(numValues)
+    writeLong(numKeyLookups)
+    writeLong(numProbes)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: do we need to carry this information after serialization?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 10]
```diff
@@ -163,29 +178,38 @@ class SQLMetricsSuite extends SparkFunSuite with SharedSQLContext {
...
-    val df = manyBytes.toSeq.toDF("a", "b").repartition(1).groupBy('a).count()
-    val metrics = getSparkPlanMetrics(df, 1, Set(2L, 0L)).get
-    Seq(metrics(2L)._2("avg hashmap probe (min, med, max)"),
-        metrics(0L)._2("avg hashmap probe (min, med, max)")).foreach { probes =>
-      probes.toString.stripPrefix("\n(").stripSuffix(")").split(", ").foreach { probe =>
-        assert(probe.toInt > 1)
+    // Assume the execution plan is:
+    // Wholestage disabled:
+    // LocalTableScan(nodeId = 4) ->Exchange (nodeId = 3) -> HashAggregate(nodeId = 2) ->
+    // Exchange(nodeId = 1) -> HashAggregate(nodeId = 0)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: can we use tree string like https://github.com/apache/spark/pull/18301/files#diff-117230dd7763cbc4f5da0be2f3a948ecR300

**Comment by cloud-fan** [comment]
LGTM, can you double check the UI and make sure it works? Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala (hunk 9 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: I think yes. In Join, we do broadcast of the hashed relation and continues to count them when querying the keys in executors.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 10 above)
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Sure.

**Comment by viirya** [comment]
Sure. I'll double check the UI later.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala (hunk 9 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: But we don't do loop up at driver side, isn't it?

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: When we build the map at driver side, we already count the numbers. Seems to me we should keep the numbers.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: oh i see

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
HashAggregate

screen shot 2017-06-29 at 10 01 57 am

HashAggregate with Limit

screen shot 2017-06-29 at 10 08 45 am

**Comment by viirya** [comment]
BroadcastHashJoin

screen shot 2017-06-29 at 10 12 27 am

**Comment by viirya** [comment]
ShuffledHashJoin

screen shot 2017-06-29 at 10 24 59 am

**Comment by viirya** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 11]
```diff
@@ -367,6 +367,22 @@ class TungstenAggregationIterator(
...
 
+  TaskContext.get().addTaskCompletionListener(_ => {
+    // At the end of the task, update the task's peak memory usage. Since we destroy
+    // the map to create the sorter, their memory usages should not overlap, so it is safe
+    // to just use the max of the two.
+    val mapMemory = hashMap.getPeakMemoryUsedBytes
+    val sorterMemory = Option(externalSorter).map(_.getPeakMemoryUsedBytes).getOrElse(0L)
+    val maxMemory = Math.max(mapMemory, sorterMemory)
+    val metrics = TaskContext.get().taskMetrics()
+    peakMemory += maxMemory
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: nit: it's more clear to call `set` here, instead of `+=`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 12]
```diff
@@ -367,6 +367,22 @@ class TungstenAggregationIterator(
...
+    spillSize += metrics.memoryBytesSpilled - spillSizeBefore
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 13]
```diff
@@ -367,6 +367,22 @@ class TungstenAggregationIterator(
...
+    metrics.incPeakExecutionMemory(maxMemory)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: not related to this PR, but `peakMemory` should pick the max memory usage among the operators in one task, instead of accumulating them?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 14]
```diff
@@ -163,29 +178,45 @@ class SQLMetricsSuite extends SparkFunSuite with SharedSQLContext {
...
+    // The executed plan looks like:
+    // HashAggregate(keys=[a#61], functions=[count(1)], output=[a#61, count#71L])
+    // +- Exchange hashpartitioning(a#61, 5)
+    //    +- HashAggregate(keys=[a#61], functions=[partial_count(1)], output=[a#61, count#76L])
+    //       +- Exchange RoundRobinPartitioning(1)
+    //          +- LocalTableScan [a#61]
+    //
+    // Assume the execution plan is:
+    // Wholestage disabled:
+    // LocalTableScan(nodeId = 4) ->Exchange (nodeId = 3) -> HashAggregate(nodeId = 2) ->
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: tree string here please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 15]
```diff
@@ -163,29 +178,45 @@ class SQLMetricsSuite extends SparkFunSuite with SharedSQLContext {
...
+    // Exchange(nodeId = 1) -> HashAggregate(nodeId = 0)
+    //
+    // Wholestage enabled:
+    // LocalTableScan(nodeId = 6) -> Exchange(nodeId = 5) ->
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala (hunk 13 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: hmm, the description of `peakExecutionMemory` in `TaskMetrics` is:

    ...The value of this accumulator should be approximately the sum of the peak sizes across all such data structures created in this task...

So it is designed to get the sum of peak memory of operators in the task. I think because the operators are not ran in sequence but in an iterator way, it's reasonable to sum the peak memory. Although the peak points of the operators might not at the same moment.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 14 above)
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: I attached the tree string. This doc is used to show `nodeId` relations.

**Comment by rxin** [comment]
hey i didn't track super closely, but it is pretty important to show at least one more digit, e.g. 1.7, rather than just 2.

**Comment by viirya** [comment]
@rxin I just revert it in previous commits. @cloud-fan should I revert it back?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala (hunk 13 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: makes sense

**Comment by cloud-fan** [comment]
@viirya  ok let's add it back

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 16]
```diff
@@ -267,10 +298,111 @@ class SQLMetricsSuite extends SparkFunSuite with SharedSQLContext {
...
+    //          +- Filter isnotnull(_1#217)
+    //             +- LocalTableScan [_1#217, _2#218]
+    //
+    // Assume the execution plan is
+    // WholeStageCodegen disabled:
+    // ... -> BroadcastHashJoin(nodeId = 1) -> Project(nodeId = 0)
+    //
+    // WholeStageCodegen enabled:
+    // ... ->
+    // WholeStageCodegen(nodeId = 0, Filter(nodeId = 4) -> Project(nodeId = 3) ->
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: can you format it a little bit? to indicate that we only have a `WholeStageCodegen`, all other plans are the inner children of `WholeStageCodegen`.

**Comment by viirya** [comment]
Verified with UI. Now showing float values with one decimal place.

screen shot 2017-06-29 at 3 10 39 pm

**Comment by viirya** [comment]
retest this please.

**Comment by cloud-fan** [comment]
LGTM, pending jenkins

**Comment by viirya** [comment]
retest this please.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by viirya** [comment]
Thanks! @cloud-fan @rxin @hvanhovell @dongjoon-hyun

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 17]
```diff
@@ -79,6 +79,11 @@ private[execution] sealed trait HashedRelation extends KnownSizeEstimation {
    * Release any used resources.
    */
   def close(): Unit
+
+  /**
+   * Returns the average number of probes per key lookup.
+   */
+  def getAverageProbesPerLookup(): Double
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: `def getAverageProbesPerLookup(): Double` -> `def getAverageProbesPerLookup: Double`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 18]
```diff
@@ -273,6 +279,8 @@ private[joins] class UnsafeHashedRelation(
   override def read(kryo: Kryo, in: Input): Unit = Utils.tryOrIOException {
     read(in.readInt, in.readLong, in.readBytes)
   }
+
+  override def getAverageProbesPerLookup(): Double = binaryMap.getAverageProbesPerLookup()
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]:
```
override def getAverageProbesPerLookup: Double = binaryMap.getAverageProbesPerLookup
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala (hunk 17 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: If you insist this change, I can do it in a related PR or a follow-up PR.

**Comment by gatorsmile** [comment]
How about getting rid of `numHashCollisions` in `BytesToBytesMap` by a follow-up PR? 

Also remove the useless `()` in the same PR

**Comment by gatorsmile** [comment]
A dumb question. Why not reporting `numHashCollisions`/`numKeyLookups`?

**Comment by viirya** [comment]
`numHashCollisions` is increased only when the hash of the key to insert is equal to the hash of an existing key in the map, but their actual keys are not equal.

But we want to show the average number of probes, which includes the cases different key hash, equal key hash (hash collision) before finding a empty slot in the map or verifying it is an existing key in the map.

**Comment by gatorsmile** [comment]
`average number of probes` will be 1 if no collision, right?

**Comment by gatorsmile** [comment]
Let me rephrase my question. Why users care the `average number of probe`, if they already know `average number of collision`?

**Comment by viirya** [comment]
My understanding is, even a key doesn't have collision when inserting, it still possible needs multiple probes to find empty slot. Personally I think number of collision doesn't tell too much information to users. It reflects how well the hash algorithm is designed.

**Comment by gatorsmile** [comment]
@viirya Could you show me the codes that need multiple probes to find an empty slot?

**Comment by viirya** [comment]
https://github.com/apache/spark/blob/fd1325522549937232f37215db53d6478f48644c/core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java#L473

In the a probe at L473, if the slot pointed by the hash code is not empty, it's possible that there's hash collision (equal hash codes, different keys), but it's possible too that the slot is occupied by a key with different hash (the if condition at L475 is false). In this case, we continue to look up an empty slot by going forward `step` at 492, and increase the number of probe.
